docs/roslyn_best_practices.md

```diff
@@ -0,0 +1,181 @@
+
+**Do**
+* Use string constants when evaluating string conditionals.
+
+**Do Not**
+* Do not use string interpolation, contenation, or format strings which construct new objects each time the analyzer is run.
+
+### 3. Enable your Analyzer to run concurrently
+Enable concurrent execution of your analyzers and prevent them from running in auto generated code.
+
+**Do**
+* Use the following in your diagnostic analyzer class.
+```cs
+    public override void Initialize(AnalysisContext context)
+    {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            ...
+    }
+```
+
+### 4. Directly share node locations with the code fixer
+If you track multiple syntax nodes in your Analyzer then you can make it easier to write your Code Fixer by using the method overload of `Diagnostic.Create` that supports [additionalLocations](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostic.create#Microsoft_CodeAnalysis_Diagnostic_Create_Microsoft_CodeAnalysis_DiagnosticDescriptor_Microsoft_CodeAnalysis_Location_System_Collections_Generic_IEnumerable_Microsoft_CodeAnalysis_Location__System_Collections_Immutable_ImmutableDictionary_System_String_System_String__System_Object___)
+
+### 5. Beware of Trivia
+When searching for patterns of code it is common to perform string comparisons. These approaches can work successfully but you should beware that trivia, the spacing we use to make code more readable, can vary from team to team. Not everyone puts a space after an assignment operator and if there are two spaces, or a tab, after the assignment operator that code is still valid.
+
+**Should Not**
+* Try to avoid `ToFullString` and solutions that require awareness of trivia.
+
+### 6. Report the diagnostic only on parts of code that would be changed by a code fixer.
+ This is a better communication to the end-user when running in Visual Studio and it makes it easier to write the Code Fixer.
+
+### 7. Do not use async methods in the synchronous context of an Analyzer. 
+If the only methods available are async, you should find another way to implement your analyzer to prevent runtime issues.
+
+**Do not**
+* Use `.Result` or `Wait()` in a Roslyn Analyzer. Forcing asynchronous code the behave synchronously can result in timing issues that are hard to debug and thread starvation.
+
+## Best Practices for Roslyn Code Fixers
+
+### 1. Code fixers handle well-known scenarios
+You may need to do a few final checks before registering your code fixer by calling `RegisterCodeFix` but you should lean into doing as many checks as possible in the analyzer.
+
+You can also accelerate your code fixer productivity by using the additionalLocations overload when reporting a diagnostic so that you can find multiple interesting locations in a document when working from a single diagnostic.
```

Comment:
nit: surround items like `additionalLocations` with a tick so it stands out as a constant concept